### ISSUE TITLE
Terraform force recreation of every servers after any Marketplace image update

### ISSUE THREAD

**Issue by jqueuniet** [terraform-providers/terraform-provider-scaleway#95]
### Terraform Version

```
Terraform v0.11.9
+ provider.scaleway v1.7.0
```

### Affected Resource(s)
* `scaleway_image`
* `scaleway_server`

### Terraform Configuration Files
```
data "scaleway_image" "ubuntu" {
    architecture    = "x86_64"
    name            = "Ubuntu Bionic"
}

resource "scaleway_server" "ns_master" {
    name = "ns0"
    image = "${data.scaleway_image.ubuntu.id}"
    state = "running"
    type = "START1-S"
    tags = ["dns", "master"]
}
```

### Debug Output
```
Terraform will perform the following actions:

-/+ scaleway_server.ns_master (new resource required)
      id:           "b7fdcf25-834d-4025-b144-878be035b293" => <computed> (forces new resource)
      boot_type:    "bootscript" => <computed>
      enable_ipv6:  "false" => "false"
      image:        "27452e61-310e-4fe5-93af-0a0bdf4c20a6" => "d4067cdc-dc9d-4810-8a26-0dae51d7df42" (forces new resource)
      name:         "ns0" => "ns0"
      private_ip:   "yyyyyyyyyyyyy" => <computed>
      public_ip:    "xxxxxxxxxxx" => "xxxxxxxxxxx"
      public_ipv6:  "" => <computed>
      state:        "running" => "running"
      state_detail: "booted" => <computed>
      tags.#:       "2" => "2"
      tags.0:       "dns" => "dns"
      tags.1:       "master" => "master"
      type:         "START1-S" => "START1-S"
```

### Expected Behavior
Instances shouldn't be trashed after image updates

### Actual Behavior
Terraform wants to replace every instance after an image update.

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`
2. wait for an image update (or force things a little by referring an older raw image ID like `27452e61-310e-4fe5-93af-0a0bdf4c20a6` for Ubuntu 18.04)
3. `terraform apply`

From what I gathered, the issue is scaleway_image only returns the last local revision ID instead of the persistent global image ID, since this is what you get in the scaleway_server structure on the other side.

I don't know if scaleway_image can store the whole history of revision IDs to compare with the ones you get from servers, or if it should be done the other way, find the persistent ID from the revision ID on servers and compare this one instead. This might get tricky depending on what Terraform Core expects from plugins.

**Comment by jqueuniet** [comment]
cc @sieben

**Comment by nicolai86** [comment]
Thanks for reporting this; I think in this case the issue is that the underlying image was recreated in the scaleway image repository: as you did not specify `most_recent` the data source would fail to compute if multiple results exists.
From your comment it seems that no error was raised, so the identity of the image changed; 

@sieben does Scaleway provide guidance on how to handle this? It seems we’re missing some kind of versioning, potentially outside of the provide image name/ hardware type.

**Comment by sblaisot** [comment]
You can workaround this problem and ignore image uuid changes with [lifecycle policy](https://www.terraform.io/docs/configuration/resources.html#lifecycle)

juste add **inside** your `scaleway_server` resource:
```
  lifecycle {
    ignore_changes = ["image"]
  }
```

Of course you will not be able to rebuild your server with a new image as long as this lines are present in your `scaleway_server` resource.

**Comment by jqueuniet** [comment]
@sblaisot Thanks for the tip, it's far better than my previous workaround of referencing raw IDs in scaleway_server structures.

@nicolai86 I'm not working on Compute directly, but I tried talking a bit about it with some people who do. From what I gathered, there is already versioning present in the backend, but it's probably not exposed in the most convenient way for this kind of use case.

I'll start with the Marketplace data structure for Ubuntu Bionic as reference (you can get this data from either https://api-marketplace.scaleway.com/images/?page=2 or https://api-marketplace.scaleway.com/images/b381b2bf-804a-4b12-91f6-9f4ff273462f )

```json
{
    "categories": [
        "distribution",
        "Ubuntu"
    ],  
    "creation_date": "2018-04-27T14:07:25.221998+00:00",
    "current_public_version": "944a4410-593f-445b-a0cb-f82621674588",
    "description": "Ubuntu is the ideal distribution for scale-out computing, Ubuntu Server helps you make the most of your infrastructure.",
    "id": "b381b2bf-804a-4b12-91f6-9f4ff273462f",
    "logo": "https://cloud.scaleway.com/images/image_logos/ubuntu.png",
    "modification_date": "2018-09-17T14:33:40.361383+00:00",
    "name": "Ubuntu Bionic",
    "organization": {
        "id": "6d6b64e5-6bad-4cc6-b7ef-2030884c3e11",
        "name": "mtouron@ocs.online.net"
    },  
    "versions": [
        {   
            "creation_date": "2018-08-31T17:20:05.484162+00:00",
            "id": "944a4410-593f-445b-a0cb-f82621674588",
            "local_images": [
                {   
                    "arch": "arm64",
                    "id": "f5c47bd9-1b97-4788-afde-dce69a6dd45e",
                    "zone": "par1"
                },  
                {   
                    "arch": "arm64",
                    "id": "a5bd979d-65ea-4a9c-ba96-0b88f0bf6b5a",
                    "zone": "ams1"
                },  
                {   
                    "arch": "x86_64",
                    "id": "d4067cdc-dc9d-4810-8a26-0dae51d7df42",
                    "zone": "par1"
                },  
                {   
                    "arch": "arm",
                    "id": "e2d81089-5a51-463c-b419-c0dfecd8c61c",
                    "zone": "ams1"
                },  
                {   
                    "arch": "arm",
                    "id": "417e38f1-b87a-404d-9f6c-63ff85e801d7",
                    "zone": "par1"
                },  
                {   
                    "arch": "x86_64",
                    "id": "6a601340-19c1-4ca7-9c1c-0704bcc9f5fe",
                    "zone": "ams1"
                }   
            ],  
            "modification_date": "2018-08-31T17:20:05.484162+00:00",
            "name": "2018-08-31T17:20:05.475847"
        }   
    ]   
}   
```

You can see three kind of IDs in there.

* `.id` is a persistent and global ID, attached to the root availability zone. It will be the same everywhere, for the whole lifetime of the Ubuntu Bionic image.
* `.versions[].id` and `.current_public_version` are the same, still a global ID, but this time referring to the revision of the image.
* `.versions[].local_images[].id` are the IDs for the real, usable images you get once an image is compiled. They are specific to each architecture and availability zone (so that make currently 6 of those, with 3 archs and 2 AZ). Those are the IDs referenced in a scaleway_server data structure.

You can browse through the Marketplace API using those IDs.

https://api-marketplace.scaleway.com/images/b381b2bf-804a-4b12-91f6-9f4ff273462f/versions/944a4410-593f-445b-a0cb-f82621674588

```json
{
    "version": {
        "creation_date": "2018-08-31T17:20:05.484162+00:00",
        "id": "944a4410-593f-445b-a0cb-f82621674588",
        "local_images": [
            {
                "arch": "arm64",
                "id": "f5c47bd9-1b97-4788-afde-dce69a6dd45e",
                "zone": "par1"
            },
            {
                "arch": "arm64",
                "id": "a5bd979d-65ea-4a9c-ba96-0b88f0bf6b5a",
                "zone": "ams1"
            },
            {
                "arch": "x86_64",
                "id": "d4067cdc-dc9d-4810-8a26-0dae51d7df42",
                "zone": "par1"
            },
            {
                "arch": "arm",
                "id": "e2d81089-5a51-463c-b419-c0dfecd8c61c",
                "zone": "ams1"
            },
            {
                "arch": "arm",
                "id": "417e38f1-b87a-404d-9f6c-63ff85e801d7",
                "zone": "par1"
            },
            {
                "arch": "x86_64",
                "id": "6a601340-19c1-4ca7-9c1c-0704bcc9f5fe",
                "zone": "ams1"
            }
        ],
        "modification_date": "2018-08-31T17:20:05.484162+00:00",
        "name": "2018-08-31T17:20:05.475847"
    }
}
```

https://api-marketplace.scaleway.com/images/b381b2bf-804a-4b12-91f6-9f4ff273462f/versions/944a4410-593f-445b-a0cb-f82621674588/local_images/d4067cdc-dc9d-4810-8a26-0dae51d7df42

```json
{
    "local_image": {
        "arch": "x86_64",
        "creation_date": "2018-08-31T17:20:05.719454+00:00",
        "id": "d4067cdc-dc9d-4810-8a26-0dae51d7df42",
        "image_versions": [
            "944a4410-593f-445b-a0cb-f82621674588"
        ],
        "modification_date": "2018-08-31T17:20:05.719454+00:00",
        "zone": "par1"
    }
}
```

What I can see as lacking from the Marketplace API would be a way to do a reverse lookup to get this permanent IDs, starting from the local revision ID. Or a way to get the whole history for an image, not only the most recent one. As things are, the API seems to only support forward lookup of version history. I opened a feature request on our side to solve this issue

Of course, all of this is only relevant to Marketplace images. User-made images are local-only and don't have any kind of versioning.

**Comment by jqueuniet** [comment]
In the end, I couldn't get to a satisfying solution. A filter mechanism was added to the `/images/`marketplace using a local image UUID as query parameter, but I fear it is too limited to completely solve the problem as it doesn't allow a real reverse lookup, and doesn't fill the gap in image version discoverability.

It might still be worth investigating, but as I no longer work for Scaleway, this is out of my hands now. I don't remember the implementation details as the marketplace change happened months ago, but @remyleone might be able to help.

**Comment by QuentinBrosse** [comment]
Hi,
We released version [v1.11.0](https://github.com/terraform-providers/terraform-provider-scaleway/blob/master/CHANGELOG.md#1110-october-01-2019) of the provider which introduces a new `scaleway_instance_server` resource.

The field [image](https://www.terraform.io/docs/providers/scaleway/r/instance_server.html#image) can contain the UUID or the label of the base image used by the server. You can use [this command](https://gist.github.com/QuentinBrosse/d8ece883ec0dbcc2dabd2d453945b4fb) to find either the right label or the right local image ID for a given commercial_type. With this new field, you can easily use a standard Scaleway image without the `scaleway_image` data source.

As you can seed in [the roadmap](https://github.com/terraform-providers/terraform-provider-scaleway/issues/125), we plan to [recode all the existing data source](https://github.com/terraform-providers/terraform-provider-scaleway/issues/225) for the `v1.13.0`.
We will take this issue into consideration to design the new `scaleway_instance_image` data source.

I will close this issue, we can continue the discussion in this one: #276.

Thanks a lot to all of you for your contribution, you rocks! 🎉
